feat(readline): add skipHistory option to rl.question() to exclude input from history

pckrishnadas88 opened the pull request:
### Description

This PR adds a new `skipHistory` boolean option to the `readline.Interface.question()` method.
When set to `true`, the user input is excluded from the readline history, useful for sensitive inputs such as passwords.

Implements the skipHistory flag throughout the readline and repl internals.

- Implements the skipHistory flag throughout the readline and repl internals.
- Defaults to false, so existing behavior is unaffected.
- Includes tests covering usage with and without the option.

### Testing

- Added parallel test` test-readline-skip-history.js` verifying correct behavior.

### Documentation
Documentation for the skipHistory option will be added in a follow-up PR or upon maintainer guidance.

**Example usage**
```js
const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

// Default behavior: input saved in history
rl.question('Enter your name: ', (name) => {
  console.log(`Hello, ${name}!`);
  console.log('History after name input:', rl.history);

  // Using skipHistory: true to avoid saving sensitive input
  rl.question('Password? ', { skipHistory: true }, (password) => {
    console.log('Password received but not stored in history.');
    console.log('History after password input (should not include password):', rl.history);
    rl.close();
  });
});

/*
Example console interaction:

Enter your name: Alice
Hello, Alice!
History after name input: [ 'Alice' ]
Password? supersecret
Password received but not stored in history.
History after password input (should not include password): [ 'Alice' ]
*/
```

Fixes: https://github.com/nodejs/node/issues/59390

<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

pckrishnadas88 committed:
readline: add skipHistory option to rl.question()

Add a boolean skipHistory option to readline.Interface.question().
When true, the input will not be saved in the readline history.
Useful for sensitive inputs like passwords.

Includes tests verifying both skipHistory enabled and disabled cases.
Docs to be added later after maintainer review.
Fixes: https://github.com/nodejs/node/issues/59390

aduh95 commented on lib/internal/repl/history.js:
```diff
@@ -44,6 +44,7 @@ const kIsFlushing = Symbol('_kIsFlushing');
 const kHistory = Symbol('_kHistory');
 const kSize = Symbol('_kSize');
 const kIndex = Symbol('_kIndex');
+const kSkipHistory = Symbol('_skipHistory');
```
Shouldn't you be reusing the same symbol instead of creating a new one?

pckrishnadas88 replied:
Will update the PR soon with the suggested changes. Thank you for the feedbacks.

pckrishnadas88 replied:
All requested changes added.

aduh95 commented on test/parallel/test-readline-skip-history.js:
```diff
@@ -0,0 +1,42 @@
+  output.isTTY = true; // simulate TTY
+  const rl = readline.createInterface({ input, output });
+
+  rl.question('Password? ', (pw) => {
```
```suggestion
  rl.question('Password? ', common.mustCall((pw) => {
```

aduh95 commented on test/parallel/test-readline-skip-history.js:
```diff
@@ -0,0 +1,42 @@
+
+  rl.question('Password? ', { skipHistory: true }, (pw) => {
+    assert.strictEqual(pw, 'secret-pass');
+    assert.strictEqual(rl.history.includes('secret-pass'), false, 
+    'Password should not be in history');
```
```suggestion
    assert.ok(!rl.history.includes('secret-pass'), 'Password should not be in history');
```

aduh95 commented on test/parallel/test-readline-skip-history.js:
```diff
@@ -0,0 +1,42 @@
+
+  rl.question('Password? ', (pw) => {
+    assert.strictEqual(pw, 'normal-input');
+    assert.strictEqual(rl.history.includes('normal-input'), true, 
+    'Input should be in history');
```
```suggestion
    assert.partialDeepStrictEqual(rl.history, ['normal-input']);
```

pckrishnadas88 committed:
readline: enhance readline skipHistory and tests

- Added common.mustCall to ensure callbacks are invoked as expected
- Used assert.partialDeepStrictEqual to verify rl.history contents precisely

pckrishnadas88 committed:
readline: fix lint errors in readline interface and symbols

- Capitalized comments per style guide
- Added missing trailing commas in objects
- Use primordials.Symbol instead of global Symbol
- Address node-core linting rules for internal modules

aduh95 commented on lib/internal/readline/symbols.js:
```diff
@@ -0,0 +1,11 @@
+const {
+  Symbol,
+} = primordials;
+const kSkipHistory = Symbol('_skipHistory');
```
Could we move this to e.g. `lib/internal/readline/utils.js` instead of creating a new file?

pckrishnadas88 replied:
I’m currently out this fix made locally but I think I need to fix coverage fails as well once everything is fixed will push again.

pckrishnadas88 committed:
readline: fix lint errors in readline interface and symbols

- Capitalized comments per style guide
- Added missing trailing commas in objects
- Use primordials.Symbol instead of global Symbol
- Address node-core linting rules for internal modules

BridgeAR reviewed (commented):
I think having such functionality makes sense.

I just wonder if we should ever save the response in the history or, if it would be best to ignore it by default. Has this ever been discussed anywhere? I couldn't find anything around that.

pckrishnadas88 commented:
> I think having such functionality makes sense.
> 
> I just wonder if we should ever save the response in the history or, if it would be best to ignore it by default. Has this ever been discussed anywhere? I couldn't find anything around that.

I'm a newcomer in this ecosystem. Saw this as an open issue so tried to do something. Issue is listed here. 

https://github.com/nodejs/node/issues/59390

pckrishnadas88 committed:
readline: fixed linting issues, test case for skipHistory false

- Address node-core linting rules for internal modules
- test case for skipHistory false

aduh95 commented:
We would want to document the new option if that's the approach we decide to go with.

pckrishnadas88 commented:
Added @nodejs/repl to the PR description since this change is REPL-related.

pckrishnadas88 commented:
> Let's just change the current implementation and release it as breaking change. I believe it was just an oversight to include these entries in the history originally. No one has likely thought about that.

Sorry I missed your reply. So what you recommend 

 1. Remove user inputs completely from history  ie no need of `skipHistory` 
 2. Make the flag `skipHistory true` by default.

pckrishnadas88 commented:
Hi @nodejs/tsc @nodejs/collaborators,

Hope you're all doing well! I wanted to gently follow up on my PR.

BridgeAR commented:
> Remove user inputs completely from history ie no need of skipHistory

Exactly. With that, there is also no need for `skipHistory` at all. Our behavior would be as if it would default to true, just without having the option.
